File: sdk/space/object-uri_test.go
package s4wave_space

import "testing"

// TestParseObjectURITable pins ParseObjectURI output. The cases must match
// sdk/space/object-uri.test.ts exactly so both runtimes parse identically.
func TestParseObjectURITable(t *testing.T) {
	cases := []struct {
		uri       string
		objectKey string
		path      string
	}{
		{"some/object/key", "some/object/key", ""},
		{"some/object/key/-/foo/bar", "some/object/key", "foo/bar"},
		{"some/object/key/-", "some/object/key", ""},
		{"key/-/", "key", ""},
		{"key/-/foo", "key", "foo"},
		{"-/foo", "foo", ""},
		{"-/-/foo", "-/foo", ""},
		{"-", "", ""},
		{"-/", "", ""},
	}
	for _, c := range cases {
		got := ParseObjectURI(c.uri)
		if got.ObjectKey != c.objectKey || got.Path != c.path {
			t.Errorf("ParseObjectURI(%q) = {%q %q}, want {%q %q}", c.uri, got.ObjectKey, got.Path, c.objectKey, c.path)
		}
	}
}

File: sdk/space/object-uri.test.ts
import { describe, expect, test } from 'vitest'
import { joinObjectUriPath, parseObjectUri } from './object-uri.js'

// The cases must match sdk/space/object-uri_test.go exactly so both runtimes
// parse identically.
describe('parseObjectUri', () => {
  const cases: Array<[string, string, string]> = [
    ['some/object/key', 'some/object/key', ''],
    ['some/object/key/-/foo/bar', 'some/object/key', 'foo/bar'],
    ['some/object/key/-', 'some/object/key', ''],
    ['key/-/', 'key', ''],
    ['key/-/foo', 'key', 'foo'],
    ['-/foo', 'foo', ''],
    ['-/-/foo', '-/foo', ''],
    ['-', '', ''],
    ['-/', '', ''],
  ]

  for (const [uri, objectKey, path] of cases) {
    test(`parses ${JSON.stringify(uri)}`, () => {
      expect(parseObjectUri(uri)).toEqual({ objectKey, path })
    })
  }
})

describe('joinObjectUriPath', () => {
  test('filters empty segments and drops a trailing dash marker', () => {
    expect(joinObjectUriPath(['a', '', 'b', '-'], false)).toBe('a/b')
  })

  test('prefixes the separator when absolute', () => {
    expect(joinObjectUriPath(['a', 'b'], true)).toBe('/a/b')
  })
})
